faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+    ----------
+    id : uuid.UUID
+        The ID of the environment.
+    steps : List[EnvironmentExecutionStep]
+        List of EnvironmentExecutionStep objects.
+    """
+
+    id = attrib()
+    steps = attrib()
+
+
+@attrs
+class EnvironmentExecutionStep(object):
+    """A single environment execution step on a server.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the environment step.
+    command : List[str]
+        The command executed by the step.
+    status : enum.Enum
+        The status of the step execution.
+    started_at : Optional[datetime]
+        The start time of the execution step.
+    finished_at : Optional[datetime]
+        The finish time of the execution step.
+    """
+
+    id = attrib()
+    command = attrib()
+    status = attrib()
+    started_at = attrib()
+    finished_at = attrib()
+
+
+@attrs
+class EnvironmentExecutionStepLog(object):
+    """The log for an environment execution step.
+    
+    Parameters
+    ----------
+    line_number : int
+        The line number within the step log.
+    content : str
+        The content of the step log.
+    """
+
+    line_number = attrib()
+    content = attrib()
+
+
+
+@attrs
+class ServerResources(object):
+    """Information about current server resource usage.
+    
+    Parameters
+    ----------
+    milli_cpus : CpuUsage
+        current CPU utilisation.
+    memory_mb : MemoryUsage
+        current memory utilisation. 
+    """
+
+    milli_cpus = attrib()
+    memory_mb = attrib()
+
+
+@attrs
+class CpuUsage(object):
+    """Current CPU usage on a server.
+    
+    Parameters
+    ----------
+    total : int
+        total CPU resource.
+    used : int
+        currently utilised CPU resource.
+    """
+
+    total = attrib()
+    used = attrib()
+
+
+@attrs
+class MemoryUsage(object):
+    """Current Memory usage on a server.
+    
+    Parameters
+    ----------
+    total : float
+        total memory resource.
+    used : float
+        current utilised memory resource.
+    cache : float
+        current cache memory usage.
+    rss: float
+        resident set size allocation. 
+    """
+
+    total = attrib()
+    used = attrib()
+    cache = attrib()
+    rss = attrib()
+
+
+class ServerAgentClient(BaseClient):
+    """Client for the Faculty server events service.
+
+    Build this client with a session directly.
+
+    >>> client = ServerAgenClient(url, session)
+
+    Parameters
+    ----------
+    url : str
+        The URL of the object storage service.
+    session : faculty.session.Session
+        The session to use to make requests.
+    """
+    def latest_environment_execution(self):
+        """Get the latest environment execution on the server.
+        
+        Returns
+        -------
+        Execution
+            The latest environment execution on a server. 
+        """
+        return self._get("/execution/latest", _ExecutionSchema())
+
+    def stream_server_events(self, endpoint):
```

Comment:
I'm increasingly thinking it doesn't make a huge amount of sense for this method and `BaseClient._stream` to be separate - maybe we can discuss at our catch up tomorrow?